STORM-2914: Implement ProcessingGuarantee.NONE in the spout instead o…

srdo opened the pull request:
…f using enable.auto.commit

https://issues.apache.org/jira/browse/STORM-2914

This change makes the spout commit offsets asynchronously when it is using ProcessingGuarantee.NONE. This behavior is very similar to how the consumer behaves when enable.auto.commit is true, but we need to do it manually due to https://issues.apache.org/jira/browse/STORM-2913. A nice side effect is that the commit timer setting in KafkaSpoutConfig now affects NONE as well as AT_LEAST_ONCE.

hmcl commented:
@srdo I left the [comment](https://issues.apache.org/jira/browse/STORM-2914?focusedCommentId=16346102&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16346102) on the JIRA.

srdo commented:
@hmcl Updated to address your list at https://issues.apache.org/jira/browse/STORM-2914?focusedCommentId=16351147&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16351147. Let me know if you're happy with the changes, and I'll squash.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:
```diff
@@ -519,6 +519,15 @@ private boolean isEmitTuple(List<Object> tuple) {
+        for (TopicPartition tp : assignedPartitions) {
+            offsetsToCommit.put(tp, new OffsetAndMetadata(kafkaConsumer.position(tp)));
+        }
+        kafkaConsumer.commitAsync(offsetsToCommit, null);
```
I would create a constant called NO_CALLBACK to improve readability

hmcl replied:
agree

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -453,37 +451,33 @@ public Builder(String bootstrapServers, Subscription subscription) {
         if (builder.kafkaProps.containsKey(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG)) {
-            throw new IllegalArgumentException("Do not set " + ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG + " manually."
-                + " Instead use KafkaSpoutConfig.Builder.setProcessingGuarantee");
+            throw new IllegalArgumentException("Do not set " + ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG
```
IllegalStateException

hmcl replied:
message should be: 

Kafka enable.auto.commit is not supported. Please set the desired ProcessingGuarantee using {@link setProcessingGuarantee}

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -453,37 +451,33 @@ public Builder(String bootstrapServers, Subscription subscription) {
     }
 
-    private static void setAutoCommitMode(Builder<?, ?> builder) {
+    private static void setPropsToFitProcessingGuarantee(Builder<?, ?> builder) {
```
setKafkaPropsForProcessingGuarantee

hmcl commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutConfigTest.java:
```diff
@@ -85,4 +90,12 @@ public void testMetricsTimeBucketSizeInSecs() {
     }
+    
+    @Test
+    public void testEnableAutoCommitIsBanned() {
```
testEnableAutoCommitNotSupported

hmcl commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutConfigTest.java:
```diff
@@ -85,4 +90,12 @@ public void testMetricsTimeBucketSizeInSecs() {
+    
+    @Test
+    public void testEnableAutoCommitIsBanned() {
+        expectedException.expect(IllegalArgumentException.class);
```
IllegalStateException.class

hmcl commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutMessagingGuaranteeTest.java:
```diff
@@ -109,7 +119,7 @@ public void testAtMostOnceModeDisregardsMaxUncommittedOffsets() throws Exception
 
     @Test
-    public void testAnyTimesModeDisregardsMaxUncommittedOffsets() throws Exception {
+    public void testNoneModeDisregardsMaxUncommittedOffsets() throws Exception {
```
testProcessingGuaranteeNoGuaranteeDisregardsMaxUncommittedOffsets

hmcl commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutMessagingGuaranteeTest.java:
```diff
@@ -153,8 +163,8 @@ public void testAtMostOnceModeCannotReplayTuples() throws Exception {
     @Test
-    public void testAnyTimesModeCannotReplayTuples() throws Exception {
-        //When tuple tracking is enabled, the spout must not replay tuples in any-times mode
+    public void testNoneModeCannotReplayTuples() throws Exception {
```
testProcessingGuaranteeNoGuaranteeCannotReplayTuples

hmcl commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutMessagingGuaranteeTest.java:
```diff
@@ -153,8 +163,8 @@ public void testAtMostOnceModeCannotReplayTuples() throws Exception {
-    public void testAnyTimesModeCannotReplayTuples() throws Exception {
-        //When tuple tracking is enabled, the spout must not replay tuples in any-times mode
+    public void testNoneModeCannotReplayTuples() throws Exception {
+        //When tuple tracking is enabled, the spout must not replay tuples in none mode
```
no-guarantee

hmcl commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutMessagingGuaranteeTest.java:
```diff
@@ -196,13 +206,37 @@ public void testAtMostOnceModeDoesNotCommitAckedTuples() throws Exception {
     @Test
-    public void testAnyTimesModeDoesNotCommitAckedTuples() throws Exception {
-        //When tuple tracking is enabled, the spout must not commit acked tuples in any-times mode because committing is managed by the consumer
+    public void testNoneModeCommitsPolledTuples() throws Exception {
```
testProcessingGuaranteeNoGuaranteeCommitsPolledTuples

hmcl commented on external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutMessagingGuaranteeTest.java:
```diff
@@ -196,13 +206,37 @@ public void testAtMostOnceModeDoesNotCommitAckedTuples() throws Exception {
-    public void testAnyTimesModeDoesNotCommitAckedTuples() throws Exception {
-        //When tuple tracking is enabled, the spout must not commit acked tuples in any-times mode because committing is managed by the consumer
+    public void testNoneModeCommitsPolledTuples() throws Exception {
+        //When using the none guarantee, the spout must commit tuples periodically, regardless of whether they've been acked
```
no-guarantee

srdo commented:
@hmcl Addressed your comments. Also fixed the storm-kafka-client docs to use release specific links, instead of linking to the 1.0.x branch. I've removed some of the listings of enum values in favor of linking to the enum javadoc as well, I don't think it's a good idea to duplicate that information. It's better if we just link to the javadoc, so the documentation doesn't get out of sync.

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -143,28 +146,28 @@ public String toString() {
+        /**
+         * The polled offsets are ready to commit immediately after being polled. The offsets are committed periodically, i.e. a message may
+         * be processed 0, 1 or more times. This behavior is similar to setting enable.auto.commit=true in the consumer, but allows the
+         * spout to control when commits occur. Commits on the defined interval
```
Commits are made asynchronously on the defined interval.

Should we also say specifically that for A_L_O and A_M_O the commits are done synchronously ?

srdo replied:
Yes, will add notes to all of them

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -453,37 +459,33 @@ public Builder(String bootstrapServers, Subscription subscription) {
+                 * error, rather than seeking to the latest (Kafka's default). This type of error will typically happen when the consumer
+                 * requests an offset that was deleted.
+                 */
+                builder.kafkaProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
```
We should print and INFO level log here saying:

LOG.info("Set Kafka property {} to {}, ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");

srdo replied:
Good point, will add the log

hmcl commented on external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java:
```diff
@@ -453,37 +459,33 @@ public Builder(String bootstrapServers, Subscription subscription) {
             }
-            builder.kafkaProps.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
         }
+        builder.kafkaProps.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
```
We should print and INFO level log here saying:

LOG.info("Set Kafka property {} to {}, ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);

srdo commented:
@hmcl Thanks for the review, caught a lot of oversights :)

hmcl commented:
+1. Let's squash and as far as I am concerned it is good to merge. Once this is squash can you please rebase STORM-2913. Thanks.
